### PR TITLE
executor,sessionctx: Add correctness for more system variables

### LINKED ISSUE #7195: Complete correctness for system variables
In #7117 we add a framework to validate input when set system variables, but not all system variables are checked in that pr because there are too much of them. We should complete these checks step by step.

What to do:
1.  Find out the specific limitation in [MySQL system variable reference](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html)
2. Add related validation in method `ValidateSetSystemVar` in `sessionctx/variable/varsutil.go`
3. Add related tests in `executor/set_test.go`

Notes:

1.  Details about MySQL system variables can be found [here](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html)
2. #7196 be be referd as an example.
3. System variables with `ScopeNone` scope need not be checked because they can't be changed dynamically.

### PR DESCRIPTION
**Pull request by spongedu**
## What is the type of the changes? (mandatory)
for #7195

- Improvement
- Compatibility

## How has this PR been tested? (mandatory)
Unittest

### CODE REVIEW

File: sessionctx/variable/varsutil.go [hunk 1]
```diff
@@ -162,6 +162,37 @@ func ValidateGetSystemVar(name string, isGlobal bool) error {
 	return nil
 }
 
+func checkIntegerSystemVar(name, value string, min int64, max uint64, vars *SessionVars) (string, error) {
```
Comment (zz-jason) [inline sessionctx/variable/varsutil.go]: can we split this function to:
- `checkUInt64SystemVar`: check whether a `uint64` value is in the range [min, max]
- `checkInt64SystemVar`: check whether a `int64` value is in the range [min, max]

File: sessionctx/variable/varsutil.go [hunk 2]
```diff
@@ -171,19 +202,10 @@ func ValidateSetSystemVar(vars *SessionVars, name string, value string) (string,
 		return value, UnknownSystemVar.GenByArgs(name)
 	}
 	switch name {
+	case ConnectTimeout:
```
Comment (jackysp) [inline sessionctx/variable/varsutil.go]: So many magic numbers in this function. Maybe we can define some constants for it.

Comment (spongedu) [reply sessionctx/variable/varsutil.go]: Most of the number literals are system variable specific and can hardly be reused, I think make them constants help little.

File: sessionctx/variable/varsutil.go (hunk 1 above)
Comment (spongedu) [reply sessionctx/variable/varsutil.go]: Most checks here range from negatives to positive integers bigger than MaxInt64. We must take both negatives and big positives together, so we can't split this function into two....

Comment (spongedu) [reply sessionctx/variable/varsutil.go]: Besides, in MySQL, system variables that with positive ranges accept negative integer, truncate them to minimum legal value and raise a warning. so we should deal with singed and unsigned together.

Comment (zz-jason) [reply sessionctx/variable/varsutil.go]: Is there a system variable ranged from negative integer to positive integer larger than `MaxInt64`?

If not, I think we can use `checkInt64SystemVar` and `checkUInt64SystemVar` to check the input parameters according to the valid value domain of the system variable. Handle the overflow error directly in these functions, which helps us to make code logic more clearer.

Comment (spongedu) [reply sessionctx/variable/varsutil.go]: ok

**Comment by spongedu** [comment]
@zz-jason  PTAL

**Comment by zz-jason** [comment]
@XuHuaiyu @jackysp PTAL

File: sessionctx/variable/varsutil.go [hunk 3]
```diff
@@ -300,6 +267,20 @@ func ValidateSetSystemVar(vars *SessionVars, name string, value string) (string,
 			return "ALL_GTIDS", nil
 		}
 		return value, ErrWrongValueForVar.GenByArgs(name, value)
+	case SQLSelectLimit:
```
Comment (zz-jason) [inline sessionctx/variable/varsutil.go]: test failed:
```
2018/08/09 22:47:40.190 main.go:693: [fatal] run test [sub_query_more] err: sql:SET SQL_SELECT_LIMIT=default;: run "SET SQL_SELECT_LIMIT=default;" at line 193 err Error 1232: Incorrect argument type to variable 'sql_select_limit'
```

**Comment by zz-jason** [comment]
@spongedu seems we should specially handle the string "default".

File: sessionctx/variable/varsutil.go [hunk 4]
```diff
@@ -25,8 +25,11 @@ import (
 	"github.com/pingcap/tidb/config"
 	"github.com/pingcap/tidb/mysql"
 	"github.com/pingcap/tidb/types"
+	"math"
```
Comment (zz-jason) [inline sessionctx/variable/varsutil.go]: please move this package up

**Comment by zz-jason** [comment]
LGTM

**Comment by zz-jason** [comment]
@jackysp PTAL

File: executor/set_test.go [hunk 5]
```diff
@@ -309,4 +309,138 @@ func (s *testSuite) TestValidateSetVar(c *C) {
 	result = tk.MustQuery("select @@time_zone;")
 	result.Check(testkit.Rows("SYSTEM"))
 
+	tk.MustExec("set @@global.max_connections=100001")
```
Comment (zhexuany) [inline executor/set_test.go]: Add comments for the test cases.

File: sessionctx/variable/varsutil.go [hunk 6]
```diff
@@ -27,6 +28,8 @@ import (
 	"github.com/pingcap/tidb/types"
 )
 
+const secondsPerYear = 31536000
```
Comment (zhexuany) [inline sessionctx/variable/varsutil.go]: Normal Year or leap year?

File: executor/set_test.go (hunk 5 above)
Comment (spongedu) [reply executor/set_test.go]: ok

File: sessionctx/variable/varsutil.go (hunk 6 above)
Comment (spongedu) [reply sessionctx/variable/varsutil.go]: MySQL don't consider leap year here. More than 31536000 seconds cause truncate warning.

Comment (zhexuany) [reply sessionctx/variable/varsutil.go]: Then 365*24*60*60 maybe much easier to understand?

Comment (spongedu) [reply sessionctx/variable/varsutil.go]: @zhexuany  I've fixed here. PTAL :)

**Pull request merged**